oci exporter: support exporting directories

jedevc opened the pull request:
Fixes https://github.com/moby/buildkit/issues/1219.

This feature adds support for specifying `unpack=true` to the oci exporter options to unpack the resulting result for the client.

This requires changes to the client to ensure that when `unpack=true` is provided, we expose a directory instead of a single file.

Because of how the containerd API is designed, the image layout file structure is only directly exposed as part of the `archiveexporter.Export` method which produces a tar - so we need to unpack it on the buildkit server before syncing it back to the client. I can't quite work out if there's a better way to do this without needing to change the containerd API.

tonistiigi reviewed (commented):
I imagined this a bit differently. Instead of unpacking and transferring files via local exporter the client side could open up a contentstore (see how `--cache-to type=local` works) and instead of setting up tarball at all the daemon side would copy the blobs over to the client-side contentstore. So there is no tarball or extract and when exporting blobs that are already available on the client side they would be just skipped.

The current implementation is just equivalent to `build | tar x` iiuc.

jedevc commented:
Have done an update to use a contentstore, the oci exporter can then use the `contentutil.CopyChain` function to copy everything over to the client properly. I've also added support for writing back an `index.json` file similar to the local cache. I think I'll come back and do a refactor of some of that code at some point, probably once #3118 merges, since the code touches similar parts of the codebase.

I've also added a commit to fix an inconsistency - we were prefixing the stores for local cache with `local:`, but weren't doing the same for OCI stores specified on the command line. I've fixed that so we prefix with `oci:` (which means we can't get any naming clashes as well) - this will need a corresponding update in buildx if we go through with that.

tonistiigi commented on client/client_test.go:
```diff
@@ -1888,7 +1889,7 @@ func testOCILayoutPlatformSource(t *testing.T, sb integration.Sandbox) {
 		},
 		OCIStores: map[string]content.Store{
-			csID: store,
+			"oci:" + csID: store,
```
I don't understand this change. The name of the store is chosen by the caller. It can be anything and can have a prefix if the client chooses it. We shouldn't arbitrarily change it. The name that caller passes in here and in LLB/Frontend-args should be identical.

If we want to make sure that these do not collide with other contentstores then this would need to be handled by the client library implementation. Eg. when calling `s.Allow(sessioncontent.NewAttachable(contentStores))` OCI stores are registered differently(either prefix or second type parameter next to the name).

tonistiigi replied:
> The local cache uses local:, 

This isn't exposed to the user. We could really just use a random identifier in there.

For the "oci-layout" the value is actually important as the user needs to use the exact same value on setting `OCIStores` and on creating the LLB definition. Also the name user chooses in there should be anything. If buildkit internally namespaces it by type field or prefix, then that should be invisible to the user/client public API.

jedevc replied:
Right! Understood, I'll add the prefixing here:

> Eg. when calling s.Allow(sessioncontent.NewAttachable(contentStores)) OCI stores are registered differently(either prefix or second type parameter next to the name).

Sorry, I understand your point now :tada:

tonistiigi commented on client/solve.go:
```diff
@@ -133,49 +137,63 @@ func (c *Client) solve(ctx context.Context, def *llb.Definition, runGateway runG
+				if err != nil {
+					return nil, err
+				}
+				contentStores["export"] = cs
```
If you want to make sure this does not collide then add a random suffix and send the full contentstore id in exporter attributes.

tonistiigi commented on client/solve.go:
```diff
@@ -133,49 +137,63 @@ func (c *Client) solve(ctx context.Context, def *llb.Definition, runGateway runG
+			}
+			switch ex.Type {
+			case ExporterOCI, ExporterDocker:
+				if err := os.MkdirAll(ex.OutputDir, 0755); err != nil {
```
Is this a different behavior from `local` where directory is created early even for broken builds(not a big issue).

Also looks like `local` defaults to `0700`? https://github.com/moby/buildkit/blob/master/session/filesync/diffcopy.go#L102

tonistiigi reviewed (commented):
Is there another name we could use here? The current name suggests that the tar is unpacked(and that this feature is much less useful). `tar=false` ?

jedevc commented:
Agreed, `tar=false` feels easier.

I think it might be worth revisiting some of the exporters at some point, especially `tar` + `local`, to see if we can merge similar behavior.

tonistiigi commented on cmd/buildctl/build/output.go:
```diff
@@ -65,19 +65,32 @@ func ParseOutput(exports []string) ([]client.ExportEntry, error) {
+	case client.ExporterTar:
+		supportFile = true
+	case client.ExporterOCI, client.ExporterDocker:
+		unpack, ok := attrs["unpack"]
```
^

tonistiigi commented:
@jedevc Note that this needs buildx update as well.
